### ISSUE TITLE
Docker for Mac `no_proxy` not honored

### ISSUE THREAD

**Issue by niothiel** [docker/for-mac#2732]
Please, check https://docs.docker.com/docker-for-mac/troubleshoot/.
Issues without logs and details cannot be debugged, and will be closed.

  - [x] I have tried with the latest version of my channel (Stable or Edge)
  - [ ] I have submitted Diagnostics (I don't believe I can, corporate restriction)
  - [ ] I have included the Diagnostics ID in this Issue (I don't believe I can, corporate restriction)

### General Introduction

This is the continuation of the cross-posted moby issue where I believe this actually belongs.

https://github.com/moby/moby/issues/36705

Since the introduction of the transparent proxy, it seems like the Docker For Mac proxy settings are
broken. If Docker For Mac is configured to have a manual proxy that includes no_proxy directives, the transparent proxy fails to obey them when any container requests the data.

I believe this is due to an inherent limitation in the way DNS name resolution works that may be an oversight in the transparent proxy implementation. If a docker container that has no proxy settings configured tries to connect to a URL like pypi.python.org, it will first resolve the DNS name to some public ip (1.2.3.4 for example) then tries to connect to that IP directly. The transparent proxy seems like it has no way of picking this up and issues a CONNECT 1.2.3.4 request to the configured proxy, even though pypi.python.org may have been configured as a proxy exemption in the Docker for Mac settings. I believe it doesn't pick this up due to it seeing only that a TCP connection to port 80/443 is being made, but it doesn't know the underlying DNS name for this operation so the Docker For Mac settings are being ignored due to lack of information (that is, the actual request the proxy sees is 1.2.3.4 instead of pypi.python.org).

This is a big limitation as the DNS names resolve to varying different IPs in a cloud-based environment, and including all possible IPs in the no_proxy variable is prohibitive. I believe the above information is accurate based on my packet-tracing efforts (specifically the transparently-proxied CONNECT {host} requests).

### Expected behavior
A `CONNECT` request being made to the proxy with the literal IP that `pypi.python.org` resolved to, even though it has been specified in the Docker for Mac settings as not using the proxy.

### Actual behavior
Observe a direct connection from the docker container to pypi.python.org via a packet-capture of the host.

### Information

```
I don't believe I can provide this information except for certain details regarding the request (I took a local capture as requested).

I don't believe this is the root cause, and is more related to transparent proxying rather than anything Mac-related, but I could be wrong.
```

### Steps to reproduce the behavior

1. Configure Docker for Mac to use a proxy server and put `pypi.python.org` into the no_proxy portion.
2. Start up a docker container with no proxy settings configured.
3. Run `curl pypi.python.org` while packet-capturing the traffic

**Comment by pgayvallet** [comment]
@djs55 ?

**Comment by djs55** [comment]
@niothiel Thanks for your very clear report.

I think your analysis is accurate; the only thing I would add is that plaintext traffic on port 80 often has a `Host:` header which we try to match against the `no_proxy` configuration. I agree though that DNS names for `no_proxy` can't work for `HTTPS` since we can't read the host header because of the encryption.

I suspect we'll have to disable the transparent proxy by default and modify the `~/docker/client.json` file instead. I'll investigate further and let you know.

**Comment by niothiel** [comment]
Great, thanks for looking into it. Let me know if I can provide any more feedback throughout your investigation.

**Comment by welchwilmerck** [comment]
Having no value for NO_PROXY seems to cause a problem for eclipse/che. As the che server container communicates with the workspace containers over http, it has to be able to connect to them directly and not be intercepted by the proxy:

```
$ docker info | grep -i proxy
HTTP Proxy: docker.for.mac.http.internal:3128
HTTPS Proxy: docker.for.mac.http.internal:3129
```
Then, a WARN from a [script](https://github.com/eclipse/che/blob/6.4.x/dockerfiles/base/scripts/base/startup_02_pre_docker.sh#L361) inside the container:

```
$ docker run -ti -v /var/run/docker.sock:/var/run/docker.sock -v ~/che20180426:/data eclipse/che start
WARN: Bound 'eclipse/che' to 'eclipse/che:6.4.0'
INFO: Proxy: HTTP_PROXY=docker.for.mac.http.internal:3128, HTTPS_PROXY=docker.for.mac.http.internal:3129, NO_PROXY=
WARN: Potential networking issue discovered!
WARN: We have identified that http and https proxies are set but no_proxy is not. This may cause fatal networking errors. Set no_proxy for your Docker daemon!
INFO: (che cli): 6.4.0 - using docker 18.03.1-ce / docker4mac
INFO: (che init): Installing configuration and bootstrap variables:
INFO: (che init):   CHE_HOST=192.168.65.3
```

```
$ ifconfig lo0
lo0: flags=8049<UP,LOOPBACK,RUNNING,MULTICAST> mtu 16384
	options=1203<RXCSUM,TXCSUM,TXSTATUS,SW_TIMESTAMP>
	inet 127.0.0.1 netmask 0xff000000
	inet6 ::1 prefixlen 128
	inet6 fe80::1%lo0 prefixlen 64 scopeid 0x1
	inet 192.168.65.2 netmask 0xffffff00
	nd6 options=201<PERFORMNUD,DAD>
```

**Comment by brymon68** [comment]
When will this be addressed? This is a major issue for those of us (I would assume many) that are behind a corporate proxy

**Comment by tonycpsu** [comment]
Yeah, it's really disappointing that this issue persists.  My dev team and I have had to stay back at `17.09.1-ce-mac42 (21090)` as a workaround.

**Comment by kastork** [comment]
Still present when running che 6.10 on 18.06.1-ce-mac74 (26766)

**Comment by leetrout** [comment]
Still an issue on Docker CE Version 2.0.0.0-beta1-mac75 (27117) Channel: edge 915f68b131 with Engine 18.09.0-ce-beta1

**Comment by derekhulley** [comment]
/lifecycle frozen
i.m.o.
Server Version: 18.06.1-ce still affected.

**Comment by primat** [comment]
What a bunch of amateurs. Why have the option if it doesn't work? Weak. Very weak.

**Comment by primat** [comment]
What a bunch of amateurs. Why have the option if it doesn't work? Weak. Very weak. Try configuring it manually, as specified here: https://docs.docker.com/network/proxy/

**Comment by djs55** [comment]
Thanks for all your patience on this. I've got a prototype fix for the HTTP proxy implementation to respect `no_proxy` via TLS SNI. If I add `pypi.python.org` to my Mac's "bypass proxy settings for these hosts & domains" then I can:
```
/ # curl https://pypi.python.org
<html><head><title>301 Moved Permanently</title></head><body><center><h1>301 Moved Permanently</h1></center></body></html>/ # 
```
while observing the following in my Desktop logs:
```
2021-02-28T16:14:11Z http-proxy time="2021-02-28T16:14:11Z" level=info msg="TLS with SNI = pypi.python.org matches no_proxy rule using a direct server connection: connected"
```
and also observing no traffic in my upstream proxy logs.

I'll let you know when I have a build suitable for testing.

**Comment by stephen-turner** [comment]
This is now fixed in Docker Desktop 3.3.0.
